### PR TITLE
persist: use protobuf for metadata encoding

### PR DESCRIPTION
**Pull request by danhhz**
We initially used Abomonation and then serde+bincode for convenience and
ease of development, but neither was ever intended as the long-term
solution for how to encode metadata.

This switches to Google's Protocol Buffers version 3 (aka [protobuf],
proto), which has a considerable number of hours of dealing with
backward and forward compatibility under its belt. Other options that
were seriously considered were Cap'n Proto and Flatbuffers.

[protobuf]: https://developers.google.com/protocol-buffers

For ease of review, this commit introduces a very thin shim that uses a
protobuf equivalent of our previously encoded structs as a middleman for
encoding and decoding. As a result, most persist code doesn't have to
change. Still TODO is figuring out whether we would like to keep this
pattern or move more toward serializing and deserializing the persist
state directly to/from the generated protobuf structs. Each has various
advantage and disadvantage tradeoffs that have yet to be fully explored.

I'm skeptical of these benchmark improvements, but they seem repeatable.

    indexed_write_drain/mem_sorted/23000000
                            time:   [142.54 ms 145.81 ms 147.92 ms]
                            thrpt:  [148.29 MiB/s 150.43 MiB/s 153.88 MiB/s]
                     change:
                            time:   [-7.8843% -5.9817% -3.9180%] (p = 0.00 < 0.05)
                            thrpt:  [+4.0778% +6.3622% +8.5591%]
                            Performance has improved.
    indexed_write_drain/mem_unsorted/23000000
                            time:   [147.02 ms 147.99 ms 149.98 ms]
                            thrpt:  [146.25 MiB/s 148.22 MiB/s 149.19 MiB/s]
                     change:
                            time:   [-6.5835% -5.1494% -3.5387%] (p = 0.00 < 0.05)
                            thrpt:  [+3.6686% +5.4290% +7.0475%]
                            Performance has improved.
    indexed_write_drain/file_sorted/23000000
                            time:   [208.09 ms 210.09 ms 212.47 ms]
                            thrpt:  [103.24 MiB/s 104.40 MiB/s 105.41 MiB/s]
                     change:
                            time:   [-2.2260% -0.0583% +2.2228%] (p = 0.96 > 0.05)
                            thrpt:  [-2.1745% +0.0583% +2.2767%]
                            No change in performance detected.
    indexed_write_drain/file_unsorted/23000000
                            time:   [209.71 ms 211.21 ms 212.55 ms]
                            thrpt:  [103.20 MiB/s 103.85 MiB/s 104.59 MiB/s]
                     change:
                            time:   [+0.1669% +2.1178% +3.6465%] (p = 0.04 < 0.05)
                            thrpt:  [-3.5182% -2.0739% -0.1666%]
                            Change within noise threshold.

### Motivation

  * This PR adds a feature that has not yet been specified.

The foundations of persist backward compatibility support.

### Tips for reviewer

If you're not familiar with protobuf, I'm happy to hop on a zoom and give you the rundown. That will also give me a good basis for what context other readers of this code may need.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [N/A] This PR adds a release note for any
  [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

**Comment by danhhz** [comment]
cc @benesch in case he has any special protobuf wisdom

**Comment by ruchirK** [comment]
will review this in 30 min

**Comment by benesch** [comment]
No special Protobuf wisdom! All the wisdom I have is already encoded in the mz_protoc crate.

File: src/persist/src/gen.rs
Comment (ruchirK) [inline src/persist/src/gen.rs:123]: while I was reviewing this I was having a hard time convincing myself whether its sufficient to just test mutating the encoded data (and keep the persisted checksum the same) vs mutating the checksum, and keeping the encoded data are the same.

I think those two cases are symmetric and check the exact same code (read the bytes, separate out old checksum and data, compute new checksum, and compare with old) but wasn't 100% sure and wanted to flag that maybe adding a second test is worth it?

File: src/persist/src/persist.proto
Comment (ruchirK) [inline src/persist/src/persist.proto:53]: I remember us talking about how you wanted to store uuids directly instead of strings at some point. did you still want to do that / would that change the proto type used here?

File: src/persist/src/persist.proto
Comment (ruchirK) [inline src/persist/src/persist.proto:18]: just for my own knowledge: whats the runtime/storage cost of using protobuf map types? is that a viable choice for these fields?

File: src/persist/src/gen.rs
Comment (ruchirK) [inline src/persist/src/gen.rs:43]: nit: stale comment

File: src/persist/src/gen.rs
Comment (ruchirK) [inline src/persist/src/gen.rs:53]: just to confirm: we don't want to keep the md5 hash as a proto field because from skimming the encoding guide it seems like theres no stable order to how fields are serialized, so its a bit chicken and egg -- you've got the encoded struct + you computed the md5 (with the md5 field blanked out to 16 0 bytes) but now theres no way to know where to place the md5 in the encoded data -- is that right?

but it seems like the encoding version could be a field in the proto spec? is that right or is there a more fundamental reason we don't want to do that? I guess for now we need a way to grab the encoding from the first byte, but that seems more like a short term hack and something we could work around

**Review by ruchirK — APPROVED** [review]
high level this looks good to me! I had some questions about why the encoding version couldn't be a field in the proto spec, but overall this all makes sense.

File: src/persist/src/gen.rs
Comment (danhhz) [reply src/persist/src/gen.rs:123]: good idea! I'll add that

File: src/persist/src/persist.proto
Comment (danhhz) [reply src/persist/src/persist.proto:18]: from https://developers.google.com/protocol-buffers/docs/overview#backwards_compatibility

a map field is the same as a repeated message with two fields (key and value). so e.g. if this was just (int64 id) -> (string name), they'd be equivalent on the wire and at runtime would be whatever the difference is between a hashmap and a vec of 2-tuples. since we also need to group the codec names in, a proto map would have a tiny bit extra of extra overhead, since the encoding would be something like `[message { key: int64, message { name, key_codec, val_codec} }]` (note the extra message). but (for better or worse) proto encoding uses varints so the overhead of this extra message in practice will be 2 bytes per entry (1 byte for field id + field type, 1 byte for message encoded length)

honestly I just didn't remember that proto maps were a thing when I did this. do you think that's worth the extra overhead?

File: src/persist/src/persist.proto
Comment (danhhz) [reply src/persist/src/persist.proto:53]: yup, i went to do it and it's a little annoying for Reasons (tm), so i held off. i was actually considering using it as test run for making a backward compatible change after this lands (and then cleaning up the cruft next time we wipe the data)

File: src/persist/src/gen.rs
Comment (danhhz) [reply src/persist/src/gen.rs:53]: > just to confirm: we don't want to keep the md5 hash as a proto field because from skimming the encoding guide it seems like theres no stable order to how fields are serialized, so its a bit chicken and egg -- you've got the encoded struct + you computed the md5 (with the md5 field blanked out to 16 0 bytes) but now theres no way to know where to place the md5 in the encoded data -- is that right?

oh that's clever. it's easy enough to hack this, the proto encoding says that if you duplicate a non-repeated field id, the last one wins. (said another way, you can "merge" two encoded protos by simply concatenating them) so we could serialize this, checksum it, and then manually encode that data and append it to the end. i really want to get this in before the next release is cut, so mind if we leave this as a TODO (might actually be a good one for you to pick off to get more familiar with protobuf)

Comment (danhhz) [reply src/persist/src/gen.rs:53]: > but it seems like the encoding version could be a field in the proto spec? is that right or is there a more fundamental reason we don't want to do that? I guess for now we need a way to grab the encoding from the first byte, but that seems more like a short term hack and something we could work around

ah damn, i tried to preemptively address this exact Q with comments but looks like I failed :). the encoding version stays outside because it's an escape hatch in case way down the line we want to stop using proto for this and use something else entirely (_extremely_ unlikely but we'll be glad to have it if it comes up)

File: src/persist/src/persist.proto
Comment (ruchirK) [reply src/persist/src/persist.proto:53]: that sounds fine to me

File: src/persist/src/persist.proto
Comment (ruchirK) [reply src/persist/src/persist.proto:18]: I think its worth it iff it means we can get rid of all of the annoying "this is a vector but really its meant to hold a set, and we need all of these complicated validations to guarantee key-uniqueness" because i find those to be the biggest pain point of keeping these as vectors. otherwise, i don't have a strong feeling

File: src/persist/src/gen.rs
Comment (ruchirK) [reply src/persist/src/gen.rs:53]: ive heard this md5 trick before, can't claim to have come up with that myself :D - no worries on doing this right away totally fine to leave it as a todo

Comment (ruchirK) [reply src/persist/src/gen.rs:53]: so, re: encoding version, just to say it back to you -- it sounds like we want _some_ identifier of what kind of data this is in a stable location, ideally near the front of the serialized data, so that down the line, we could use this identifier + location to determine what kind of data we're looking at, and the proto spec doesn't give us a way to express "I'd like encoded bytes [start..end] to be foo please" and so we have to handroll?

If so, makes sense and I guess its the best we can do. I'll defer to your judgement here because. The small downside i noticed is that it seems like we're currently encoding into a vec, and then copying its contents into another buffer, which is a small amount of extra copying, but im sure theres ways to work around that,

File: src/persist/src/gen.rs
Comment (danhhz) [reply src/persist/src/gen.rs]: Done

File: src/persist/src/gen.rs
Comment (danhhz) [reply src/persist/src/gen.rs:53]: yeah that's right. good point on the temp vec, even without pushing the checksum into proto, that's not necessary if we put the checksum at the end, so i did that. i tried to hack in actually computing the checksum as we go and it was _just_ hairy enough that I left that part as a TODO, but it won't be a breaking change

File: src/persist/src/persist.proto
Comment (danhhz) [reply src/persist/src/persist.proto:19]: okay, I think this all depends on how the validation stuff shakes out and whether we keep stuff like ArrangementMeta around or switch to directly constructing things from protos. we'll have more breaking changes before this settles down (plus it's possible to do without a breaking change), so left it as a TODO

**Review by danhhz — COMMENTED** [review]
TFTR!

File: src/persist/src/persist.proto
Comment (danhhz) [reply src/persist/src/persist.proto:19]: haha jk. convinced myself we should do this now

Comment (danhhz) [reply src/persist/src/persist.proto]: Done

File: src/persist/src/indexed/encoding.rs
Comment (ruchirK) [inline src/persist/src/indexed/encoding.rs:644]: I think you need a stable order so that you can have stable comparisons with previously serialized metadata (without having to write a specialized comparator that sorts the keys each time you want to compare)

Comment (danhhz) [reply src/persist/src/indexed/encoding.rs:644]: yeah that's exactly why I had to sort it. turns out I meant BTreeMap

**Pull request merged**
